src/coreclr/nativeaot/System.Private.CoreLib/src/System/CrashInfo.cs

```diff
@@ -0,0 +1,390 @@
+            {
+                if (!OpenValue("inner"u8, '['))
+                    return false;
+
+                // Write as many inner exceptions that will fit
+                if (aggregate is not null)
+                {
+                    foreach (Exception ex in aggregate.InnerExceptions)
+                    {
+                        if (!WriteExceptionWithFallback(default, ex, maxMessageSize, maxNumberStackFrames, maxMethodNameSize))
+                            break;
+                    }
+                }
+                else
+                {
+                    WriteExceptionWithFallback(default, exception.InnerException, maxMessageSize, maxNumberStackFrames, maxMethodNameSize);
+                }
+
+                CloseValue(']');
+            }
+            CloseValue('}');
+            return true;
+        }
+
+        /// <summary>
+        /// Write an exception stack frame to the triage buffer
+        /// </summary>
+        /// <param name="frame">the stack frame instance to write</param>
+        /// <param name="maxNameSize">limits the size of the frame type name</param>
+        /// <returns>true - success, false - out of triage buffer space</returns>
+        private bool WriteStackFrame(StackFrame frame, int maxNameSize)
+        {
+            if (!OpenValue(default, '{'))
+                return false;
+
+            nint ip = frame.GetNativeIPAddress();
+            if (!WriteHexValue("ip"u8, (ulong)ip))
+                return false;
+
+            if (!WriteHexValue("offset"u8, frame.GetNativeOffset()))
+                return false;
+
+            string method = DeveloperExperience.GetMethodName(ip, out IntPtr _);
+            if (method != null)
+            {
+                if (!WriteValue("name"u8, method, maxNameSize))
+                    return false;
+            }
+            CloseValue('}');
+            return true;
+        }
+
+        private bool WriteHexValue(ReadOnlySpan<byte> key, ulong value) => WriteValue(key, $"0x{value:X}".AsSpan());
+
+        private bool WriteHexValue(ReadOnlySpan<byte> key, int value) => WriteValue(key, $"0x{value:X}".AsSpan());
+
+        private bool WriteValue(ReadOnlySpan<byte> key, int value) => WriteValue(key, $"{value}".AsSpan());
+
+        private bool WriteValue(ReadOnlySpan<byte> key, string value, int max = int.MaxValue)
+        {
+            if (!OpenValue(key, '"'))
+                return false;
+
+            // Escape the special JSON characters.
+            int count = 0;
+            foreach (char c in value)
+            {
+                // Have we reached the max number of chars?
+                if (++count > max)
+                    break;
+
+                if (char.IsControl(c) || c == '"' || c == '\\')
+                {
+                    if (!WriteChars($"\\u{((ushort)c):X4}".AsSpan()))
+                        return false;
+                }
```

Comment:
```suggestion
                if (c < 0x20)
                {
                    if (!WriteChars($"\\u{((ushort)c):X4}".AsSpan()))
                        return false;
                }
                else if (c == '"')
                {
                    if (!WriteBytes(u8"\\\""))
                        return false;
                }
                else if (c == '\\')
                {
                    if (!WriteBytes(u8"\\\\"))
                        return false;
                }
```
`char.IsControl` does not match what needs to be escaped in json. Also, we can do more natural escaping for `"` and `\` since we are checking for them explicitly anyway and they are going to be somewhat common in exception messages.

Reply:
Thanks. That is a lot better.